node/pkg/watchers/aptos/watcher.go

```diff
@@ -100,17 +121,181 @@ func (e *Watcher) Run(ctx context.Context) error {
 		zap.String("healthQuery", aptosHealth),
 	)
 
-	// the events have sequence numbers associated with them in the aptos API
-	// (NOTE: this is not the same as the wormhole sequence id). The event
-	// endpoint is paginated, so we use this variable to keep track of which
-	// sequence number to look up next.
+	// Important: There are 3 different sequence/version numbers in play:
+	// 1. Aptos Version: Transaction version number (like block number in EVM)
+	// 2. Aptos Sequence Number: Event-specific counter for WormholeMessage events
+	// 3. Wormhole Sequence: Protocol sequence inside the event data (goes into VAA)
+	//
+	// nextSequence tracks the Aptos Sequence Number (not version, not wormhole seq)
 	var nextSequence uint64 = 0
 
 	timer := time.NewTicker(time.Second * 1)
 	defer timer.Stop()
 
 	supervisor.Signal(ctx, supervisor.SignalHealthy)
 
+	if e.useIndexer {
+		return e.runIndexerMode(ctx, logger, nextSequence)
+	} else {
+		return e.runLegacyMode(ctx, logger, nextSequence)
+	}
+}
+
+func (e *Watcher) runLegacyMode(ctx context.Context, logger *zap.Logger, nextSequence uint64) error {
+	// Legacy REST API implementation from original code
+	timer := time.NewTicker(time.Second * 1)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case r := <-e.obsvReqC:
+			if r.ChainId > math.MaxUint16 || vaa.ChainID(r.ChainId) != e.chainID {
+				panic("invalid chain ID")
+			}
+
+			// uint64 will read the *first* 8 bytes, but the sequence is stored in the *last* 8.
+			nativeSeq := binary.BigEndian.Uint64(r.TxHash[24:])
+
+			logger.Info("Received obsv request", zap.Uint64("tx_hash", nativeSeq))
+
+			// Legacy mode: Use REST API to get events by sequence
+			eventsQuery := fmt.Sprintf(`%s/v1/accounts/%s/events/%s?limit=1&start=%d`, e.aptosRPC, e.aptosAccount, e.aptosHandle, nativeSeq)
+			eventsJson, err := e.retrievePayload(eventsQuery)
+			if err != nil {
+				logger.Error("retrievePayload", zap.Error(err))
+				p2p.DefaultRegistry.AddErrorCount(e.chainID, 1)
+				continue
+			}
+
+			if !gjson.Valid(string(eventsJson)) {
+				logger.Error("InvalidJson: " + string(eventsJson))
+				p2p.DefaultRegistry.AddErrorCount(e.chainID, 1)
+				continue
+			}
+
+			events := gjson.ParseBytes(eventsJson)
+			for _, event := range events.Array() {
```

Comment:
Might be a good idea to do a length check and log on cases where the events.Array is empty. Otherwise the loop will be skipped silently

Reply:
Added